Use exec instead of forked child process for JVM

ept opened the pull request:
Using os.system() for creating the JVM forks a child of the Python
process. This makes it difficult for process managers like daemontools
or runit to kill/restart the process at admins' command: the process
manager tracks the PID of the Python process instead of that of the JVM;
and terminating the Python process unfortunately doesn't kill its JVM
child.

Some process managers let you configure the number of times the service
forks, or try to detect forks, but it's a bit of a rabbit-hole. Better
to make the service not fork in the first place.

For that reason I've replaced the call to os.system() with os.execvp(),
which replaces the Python process with the JVM, and thus lets the
process manager track the right PID.

os.execvp() requires that you give command-line arguments as an array of
strings, rather than one long whitespace-separated string, so I had to
change some of the surrounding code accordingly. Note this has the
pleasant side-effect that Storm should now also run if you install it in
a directory that has spaces in its path name.

ept committed:
Use exec instead of forked child process for JVM

Using os.system() for creating the JVM forks a child of the Python
process. This makes it difficult for process managers like daemontools
or runit to kill/restart the process at admins' command: the process
manager tracks the PID of the Python process instead of that of the JVM;
and terminating the Python process unfortunately doesn't kill its JVM
child.

Some process managers let you configure the number of times the service
forks, or try to detect forks, but it's a bit of a rabbit-hole. Better
to make the service not fork in the first place.

For that reason I've replaced the call to os.system() with os.execvp(),
which replaces the Python process with the JVM, and thus lets the
process manager track the right PID.

os.execvp() requires that you give command-line arguments as an array of
strings, rather than one long whitespace-separated string, so I had to
change some of the surrounding code accordingly. Note this has the
pleasant side-effect that Storm should now also run if you install it in
a directory that has spaces in its path name.

nathanmarz commented on bin/storm:
```diff
@@ -208,12 +211,15 @@ def nimbus(klass="backtype.storm.daemon.nimbus"):
     cppaths = [STORM_DIR + "/log4j", STORM_DIR + "/conf"]
-    childopts = confvalue("nimbus.childopts", cppaths) + " -Dlogfile.name=nimbus.log -Dlog4j.configuration=storm.log.properties"
+    jvmopts = confvalue("nimbus.childopts", cppaths).split() + [
+        "-Dlogfile.name=nimbus.log",
```
I don't think this is quite right, since the childopts could be something like -Dsome.config="aaa bbb".

nathanmarz commented on bin/storm:
```diff
@@ -225,12 +231,15 @@ def supervisor(klass="backtype.storm.daemon.supervisor"):
     """
     cppaths = [STORM_DIR + "/log4j", STORM_DIR + "/conf"]
-    childopts = confvalue("supervisor.childopts", cppaths) + " -Dlogfile.name=supervisor.log -Dlog4j.configuration=storm.log.properties"
+    jvmopts = confvalue("supervisor.childopts", cppaths).split() + [
```
Same as above. Should probably make this parsing piece its own function.

nathanmarz commented on bin/storm:
```diff
@@ -243,11 +252,14 @@ def ui():
     cppaths = [STORM_DIR + "/log4j", STORM_DIR + "/conf"]
-    childopts = confvalue("ui.childopts", cppaths) + " -Dlogfile.name=ui.log -Dlog4j.configuration=storm.log.properties"
+    jvmopts = confvalue("ui.childopts", cppaths).split() + [
+        "-Dlogfile.name=ui.log",
```
Same as above.

nathanmarz commented:
Left a few comments on the pull request. Also, can you send me a contributor's agreement? Instructions at the bottom of this page: https://github.com/nathanmarz/storm/wiki/Contributing-to-Storm

jshuping commented:
Thanks guys, I ran into the same issue and was thinking of a similar patch!  Just need to do something about the end of shell()  (~line 194) as this change to using execvp() will mean that last os.system() in shell() will no longer get called.

ept committed:
Allow quoted whitespace in childopts

For example: '-Dsome.config="aaa bbb" -Dother.config=ccc'
gets parsed into ['-Dsome.config=aaa bbb', '-Dother.config=ccc']
-- like what bash does.

Sorry for the heavy regexes. I think pyparsing has a built-in parser for
this, but I'd rather not add another dependency for this fairly minor
feature. I don't think there's anything suitable in the Python stdlib.

ept committed:
Don't forget to delete Storm shell's temp jar

Need to use a subprocess, not exec, since exec never returns and thus
the "rm" never gets executed.

ept commented:
@nathanmarz No problem, I've added a proper arguments parser: 0749adbe72360e38d82f33392b339977e472b9de. Contributor agreement coming shortly.

@jshuping Nicely spotted, thanks. Fixed: 08224b2a8a27ea6b5799bbd339da64662eda1b2a

shrijeet commented:
@ept  I tried your patch on 0.7.2 and got http://pastie.org/4076250 , I printed type() of various entities being concatenated in exec_storm_class method and found args to be of type tuple.  following change fixes it : 

``` python
] + jvmopts + [klass] + list(args)
```

I was running "storm list" when I first encountered this. Python version is 2.6

shrijeet commented:
@ept excuse me if I am wrong, python is not my forte.

ept committed:
Fix regression with passing args through to JVM

Python's *args in a method definition creates a tuple, not a list; and
tuples can't be concatenated with lists.

ept commented:
@shrijeet Thanks — I had tested running Nimbus, supervisors etc. but hadn't tested the client-side commands. Your fix is good, I've added a commit for it to this branch.

@nathanmarz I think it's all stable now. Could you give it another look?

ept committed:
Make docstring raw, to preserve escaping examples

ept committed:
Handle confvalue returning None

ept committed:
Make confvalue() never return None

There are three places in the code where the return value of confvalue()
is concatenated with a string, which fails if the value is None. Rather
than changing them all, I thought it would be better to make confvalue()
return an empty string if the configuration for that value is absent.

dctrwatson commented:
We use runit and having this issue for now I applied this directly, but would be nice to have it in mainline

ept commented:
We've also been using this patched version successfully on Linux and Mac OS. My only worry about `execvp` is that it may not work on Windows. The Python docs say it does, but it would be good to test it. Does anyone have a Windows machine handy on which they can test it?

nathanmarz commented on bin/storm:
```diff
@@ -82,12 +84,37 @@ def print_remoteconfvalue(name):
+    args = [re.compile(r"'((?:[^'\\]|\\.)*)'").sub('\\1', x) for x in args]
+    return [re.compile(r'\\(.)').sub('\\1', x) for x in args]
+
+def exec_storm_class(klass, jvmtype="-server", jvmopts=[], extrajars=[], args=[], fork=False):
```
We need to add fork=True for all the Storm commands, right? Maybe it would be better for fork to default to True, since the number of commands is likely to grow faster than the number of daemons.

ept replied:
@nathanmarz  `fork=True` is only needed if the script needs to do something after the call to `exec_storm_class`. Currently, `shell` is the only command for which this is the case (it needs to remove the temporary jar file). For the other client commands, either way should be fine.

nathanmarz replied:
Ah, I understand now. Thanks for clarifying.

nathanmarz commented:
Merged, thanks!

ept commented:
Awesome, thanks Nathan!
